### ISSUE TITLE
std::array json serialization extremely verbose

### ISSUE THREAD

**Issue by patrikhuber** [USCiLab/cereal#288]
Hi,

When serialising for example an `std::array<int, 3> tri { 1, 2, 3 }`, the generated json is extremely verbose:

```
"value0": {
    "value0": 1,
    "value1": 2,
    "value2": 3
}
```

The values themselves are only 1 character, while the name `value0` takes up 6 characters, plus the additional syntax elements. Even when the value is for example a float, the space the identifiers take still nearly double the required space.

This is a big issue when you serialize something like a `std::vector<std::array<int, 3>> triangle_list`, and this list contains a few ten-thousand elements.

I could probably work around this by creating a custom class holding three triangle indices with its own `serialize()` function, but this would require me to pass around this custom class everywhere in my API instead of a simple `std::array<int, 3>` that communicates its intent to the user much more directly - everybody knows at first glance what the type `std::array<int, 3>` is, while if it's a custom class, it creates an additional layer of abstraction and the user has to read the documentation to fully get what's going on.

Can't we serialize `std::array`'s in a better way to create less verbosity in general?

**Comment by stevehickman** [comment]
I second that.

An array should be turned into a JSON array not a map, so the example below should be something like:

“tri”: [ 1, 2, 3]

Regards,

Steve H.

**Comment by reuk** [comment]
You could just replace the array serialization routine with something like this (untested but whatever):

template <typename Archive, typename T, size_t S>
void serialize(Archive& archive, std::array<T, S>& m) {
    cereal::size_type s = S;
    archive(cereal::make_size_tag(s));
    if (s != S) {
        throw std::runtime_error("array has incorrect length");
    }
    std::for_each(m.begin(), m.end(), [&archive](auto& i) { archive(i); });
}

**Comment by patrikhuber** [comment]
@stevehickman: Your reply convinced me even more that I would actually consider this a "bug" in cereal, an `std::array` should definitely be serialized into a JSON array. I would love to hear @AzothAmmo's opinion on it and whether there was any reason for the current, kind of weird implementation, since an array is really not a map. And of course whether we can change it :-)

@reuk: Thanks, I haven't tried the code yet, but the [current code](https://github.com/USCiLab/cereal/blob/master/include/cereal/types/array.hpp#L64-L65) has the same `for`-loop, I suspect your code outputs the same JSON?

**Comment by reuk** [comment]
I believe that `make_size_tag` will suppress the 'value0, value1 etc.' behaviour and serialize to an array rather than an object.

[docs](http://uscilab.github.io/cereal/assets/doxygen/group__Utility.html#ga4d4b256b14671c8270d06f40e4c4a077)

**Comment by patrikhuber** [comment]
@reuk: Yes, you're right! I tested your code (it works without modifications), and indeed, `make_size_tag` suppresses generating a map. With your code, the generated JSON is:

```
    "value0": [
        1,
        2,
        3
    ]
```

* @AzothAmmo how do you feel about a PR to `cereal/types/array.hpp` with this change?

* In the meantime, is there a way I can make this change locally, without modifying cereal headers? How would I make cereal take my/@reuk's implementation over the ones in `cereal/types/array.hpp`?

**Comment by reuk** [comment]
Think a pull request with this change is a bad idea, because it breaks deserialization for files which have been serialized with older library versions.

Re: using my impl over the ones in the library, it should be as simple as putting my impl in a header, and then just not including `cereal/types/array.hpp`, right?

**Comment by stevehickman** [comment]
If this is going to break backwards compatibility, is there an existing or preferred mechanism to select at compile time which implementation to use? Obviously macros are an approach.  Traits might be another.

Any preference?

Regards,

Steve H.

**Comment by patrikhuber** [comment]
The cereal documentation explicitly states that there is no guarantee for archives to work between versions! So I think this is something that could be committed to `develop` and then merged with the next release, 1.2 or so for example. I don't see an issue with breaking it to improve the library, given everybody knows that when using cereal (this is for example one of the differences to boost serialization).

**Comment by M2tM** [comment]
So long as breaking changes are work-around able I'm happy.  For this instance one could (with a version check) conditionally load the old values into a map and manually unpack them into an array.  That would not be the end of the world.

**Comment by AzothAmmo** [comment]
Breaking changes are fine as long as they are justified. I agree that we're probably doing the wrong thing here, especially given how we serialize vector. However, `cereal/types/array.hpp` is a generic implementation and adding a `size_tag` is not a no-op for other archives, so it may not be the best overall solution. We can also use the prologue/epilogue functionality to do what `SizeTag` is doing by calling `makeArray` in the JSON archive, though this would require including `cereal/types/array.hpp` in `json.hpp`. Have to compromise somewhere.

@patrikhuber You can implement this locally by specializing the archive template parameter of the save/load functions for `std::array`, e.g.:

```cpp
template <class T, class N> inline
void /* might need enable_if copied */ load( cereal::JSONOutputArchive & ar, std::array<T, N> & array )
{
  cereal::size_type size;
  ar( cereal::make_size_tag( size ) );
  // etc
}
```

If this is going to break things I'll slate it for 1.3 since 1.2 is mostly new features or bug fixes. 1.2 has essentially been done for a while but I need to clean up the last two Windows related issues.
